File: GameServer/Game.py

class Game:
    def __init__(self, name, nb_players, size, nb_alignments_max):
    	self.name = name
    	self.nb_players = nb_players
    	self.size = size
    	self.nb_alignments_max = nb_alignments_max
    	self.size = size
    	self.grid = []
    	self.players = []
    	self.waiting = True
    	self.admin = None
        self.nb_alignments = 0

    def vertical_check(x, y, player):
        j1 = y
        j2 = y

        while j1 > 0 && (self.grid[x+self.size*(j1-1)].state == 1):
            j1--

        while j2 < (self.size-1) && (self.grid[x+self.size*(j2+1)].state == 1):
            j2++


        if (j2-j1+1) >= 5:
            for j in range (j1, j1+5):
                square = self.grid[x+self.size*j]
                square.state = 2
                square.player = player

    def horizontal_check(x, y, player):
        i1 = x
        i2 = x

        while i1 > 0 && (self.grid[(i1-1)+self.size*y].state == 1):
            i1--

        while i2 < (self.size-1) && (self.grid[(i2+1)+self.size*y].state == 1):
            i2++


        if (i2-i1+1) >= 5:
            for i in range (i1, i1+5):
                square = self.grid[i+self.size*y]
                square.state = 2
                square.player = player


    def diag_right_check(x, y, player):
        i1, j1 = x, y
        i2, j2 = x, y

        while i1 < (self.size-1) && j1 > 0 && (self.grid[(i1+1)+self.size*(j1-1)].state == 1):
            i1++
            j1--

        while i2 > 0 && j2 < (self.size-1) && (self.grid[(i2-1)+self.size*(j2+1)].state == 1):
            i2--
            j2++


        if (i1-i2+1) >= 5:
            for j in range (j1, j1+5):
                square = self.grid[i1+self.size*j]
                square.state = 2
                square.player = player
                i1--

    def diag_left_check(x, y, player):
        i1, j1 = x, y
        i2, j2 = x, y

        while i1 > 0 && j1 > 0 && (self.grid[(i1-1)+self.size*(j1-1)].state == 1):
            i1--
            j1--

        while i2 < (self.size-1) && j2 > (self.size-1) && (self.grid[(i2+1)+self.size*(j2+1)].state == 1):
            i2++
            j2++


        if (i2-i1+1) >= 5:
            for i in range (i1, i1+5):
                square = self.grid[i+self.size*j1]
                square.state = 2
                square.player = player
                j1++


    def check(x, y, player):
        self.horizontal_check(x, y, player);
        self.vertical_check(x, y, player);
        self.diag_left_check(x, y, player);
        self.diag_right_check(x, y, player);

    def play(x, y, player):
        #On vérifie que x et y sont des cases existantes
        if (x<0 or x>self.size-1) or (y<0 or y>self.size-1):
            return 1

        square = self.grid[x+self.size*y]
        if square.state == 0:
            square.state = 1
            square.player = player

        self.check()

        return 0

File: Client/Game.cpp
#include "Game.hpp"
#include <QDebug>

Game::Game(std::string name, int nbPlayers, int size, int nbAlignmentsMax, std::string admin):
    m_name(name),
    m_nbPlayers(nbPlayers),
    m_size(size),
    m_nbAlignmentsMax(nbAlignmentsMax),
    m_admin(admin)

{
    for(int j = 0; j < size; j++)
    {
        for(int i = 0; i < size; i++)
        {
            Square* square = new Square();
            square->setPosition(i, j);
            m_grid.push_back(square);
        }
    }
}

Game::~Game(void)
{

}

std::string Game::getName()
{
    return m_name;
}

int Game::getNbPlayers()
{
    return m_nbPlayers;
}

int Game::getSize(void)
{
    return m_size;
}

int Game::getNbAlignments()
{
    return m_nbAlignments;
}

std::string Game::getAdmin()
{
    return m_admin;
}

//Pas terrible
Square* Game::getSquare(int x, int y)
{
    return m_grid[x+m_size*y];
}

void Game::addPlayer(QString player)
{

}

void Game::play(int x, int y, std::string player)
{
    Square* square = m_grid[x+m_size*y];
    if(square->getState() == 0)
    {
        square->setState(1);
        check(x, y, player);
    }
}

void Game::diagLeftCheck(int x, int y, std::string player)
{
    //On remonte la diagonale vers la gauche
    int i1 = x, j1 = y;
    int i2 = x, j2 = y;
    while(i1 > 0 && j1 > 0 && (m_grid[(i1-1)+m_size*(j1-1)]->getState() == 1))
    {
        i1--;
        j1--;
    }

    while(i2 < (m_size-1) && j2 < (m_size-1) && (m_grid[(i2+1)+m_size*(j2+1)]->getState() == 1))
    {
        i2++;
        j2++;
    }

    if((i2-i1+1) >= 5)
    {
        for(int i = i1; i<(i1+5); i++)
        {
            Square* square = m_grid[i+m_size*j1];
            square->setState(2);
            square->setPlayer(player);
            j1++;
        }
    }
}

void Game::diagRightCheck(int x, int y, std::string player)
{
    //On remonte la diagonale vers la droite
    int i1 = x, j1 = y;
    int i2 = x, j2 = y;
    while(i1 < (m_size-1) && j1 > 0 && (m_grid[(i1+1)+m_size*(j1-1)]->getState() == 1))
    {
        i1++;
        j1--;
    }

    while(i2 > 0 && j2 < (m_size-1) && (m_grid[(i2-1)+m_size*(j2+1)]->getState() == 1))
    {
        i2--;
        j2++;
    }

    if((i1-i2+1) >= 5)
    {
        for(int j = j1; j<(j1+5); j++)
        {
            Square* square = m_grid[i1+m_size*j];
            square->setState(2);
            square->setPlayer(player);
            i1--;
        }
    }
}

void Game::horizontalCheck(int x, int y, std::string player)
{
    //On remonte la diagonale vers la droite
    int i1 = x;
    int i2 = x;

    while(i1 > 0 && (m_grid[(i1-1)+m_size*y]->getState() == 1))
    {
        i1--;
    }

    while(i2 < (m_size-1) && (m_grid[(i2+1)+m_size*y]->getState() == 1))
    {
        i2++;
    }


    if((i2-i1+1) >= 5)
    {
        for(int i = i1; i<(i1+5); i++)
        {
            Square* square = m_grid[i+m_size*y];
            square->setState(2);
            square->setPlayer(player);
        }
    }
}

void Game::verticalCheck(int x, int y, std::string player)
{
    int j1 = y;
    int j2 = y;

    while(j1 > 0 && (m_grid[x+m_size*(j1-1)]->getState() == 1))
    {
        j1--;
    }

    while(j2 < (m_size-1) && (m_grid[x+m_size*(j2+1)]->getState() == 1))
    {
        j2++;
    }


    if((j2-j1+1) >= 5)
    {
        for(int j = j1; j<(j1+5); j++)
        {
            Square* square = m_grid[x+m_size*j];
            square->setState(2);
            square->setPlayer(player);
        }
    }
}

void Game::check(int x, int y, std::string player)
{
    horizontalCheck(x, y, player);
    verticalCheck(x, y, player);
    diagLeftCheck(x, y, player);
    diagRightCheck(x, y, player);
}

